hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/snappy/SnappyDecompressor.java

```diff
@@ -276,13 +258,27 @@ public void end() {
     // do nothing
   }
 
-  private native static void initIDs();
+  private int decompressBytesDirect() throws IOException {
+    if (compressedDirectBufLen == 0) {
+      return 0;
+    } else {
+      // Set the position and limit of `compressedDirectBuf` for reading
+      compressedDirectBuf.position(0).limit(compressedDirectBufLen);
+      // There is compressed input, decompress it now.
+      int size = Snappy.uncompressedLength((ByteBuffer) compressedDirectBuf);
+      if (size > uncompressedDirectBuf.capacity()) {
```

Comment:
Should we check with `uncompressedDirectBuf.remaining` instead of `capacity`?

Reply:
Before calling `decompressBytesDirect`, we always reset `uncompressedDirectBuf`. So both method returns the same result. But you are right, I think .remaining is better.

Reply:
`decompressDirect` also calls `decompressBytesDirect`, but the `uncompressedDirectBuf` is passed in as an argument. I think it has danger to assume `uncompressedDirectBuf` is always reset.